pkg/kubectl/cmd/util/conversion.go

```diff
@@ -0,0 +1,37 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/kubernetes/pkg/api/legacyscheme"
+)
+
+// AsDefaultVersionedOrOriginal returns the object as a Go object in the external form if possible (matching the
+// group version kind of the mapping, or obj if it cannot be converted.
+// Deprecated this function will be removed once calling code is updated to indicate the correct
+// negoticatedserializers during construction of the builder
+func AsDefaultVersionedOrOriginal(obj runtime.Object) runtime.Object {
+	converter := runtime.ObjectConvertor(legacyscheme.Scheme)
+	groupVersioner := runtime.GroupVersioner(schema.GroupVersions(legacyscheme.Registry.RegisteredGroupVersions()))
```

Comment:
this seems like it might break on sparse API groups where the incoming object does not exist in the preferred group (e.g. cronjobs, which do not exist in batch/v1)... the previous code used the GVK the object was actually retrieved as to determine the destination GroupVersion, and this code doesn't.

Reply:
> this seems like it might break on sparse API groups where the incoming object does not exist in the preferred group (e.g. cronjobs, which do not exist in batch/v1)... the previous code used the GVK the object was actually retrieved as to determine the destination GroupVersion, and this code doesn't.

Updated. turns out we're contained to just the set commands (yay), and I made it prefer the mapping for now.  A few more pulls and this will disappear and we'll simply set the right scheme to begin with (not the current state).